e2e/tests/transfer/forwarding_test.go

```diff
@@ -108,3 +109,99 @@ func (s *TransferForwardingTestSuite) testForwardingThreeChains(lastChainVersion
 		s.Require().Equal(expected, actualBalance.Int64())
 	})
 }
+
+// TestForwardingWithUnwindSucceeds tests the forwarding scenario in which
+// a packet is sent from A to B, then unwound back to A and forwarded to C
+// The overall flow of the packet is:
+// A ---> B
+// B --(unwind)-->A --(forward)-->B --(forward)--> C
+func (s *TransferForwardingTestSuite) TestForwardingWithUnwindSucceeds() {
+	t := s.T()
+	ctx := context.TODO()
+	relayer, chains := s.GetRelayer(), s.GetAllChains()
+
+	chainA, chainB, chainC := chains[0], chains[1], chains[2]
+
+	channelAtoB := s.GetChainAChannel()
+	channelBtoC := s.GetChainChannel(testsuite.ChainChannelPair{ChainIdx: 1, ChannelIdx: 1})
+
+	chainAWallet := s.CreateUserOnChainA(ctx, testvalues.StartingTokenAmount)
+	chainAAddress := chainAWallet.FormattedAddress()
+	chainADenom := chainA.Config().Denom
+
+	chainBWallet := s.CreateUserOnChainB(ctx, testvalues.StartingTokenAmount)
+	chainBAddress := chainBWallet.FormattedAddress()
+
+	chainCWallet := s.CreateUserOnChainC(ctx, testvalues.StartingTokenAmount)
+	chainCAddress := chainCWallet.FormattedAddress()
+
+	t.Run("IBC transfer from A to B", func(t *testing.T) {
+		inFiveMinutes := time.Now().Add(5 * time.Minute).UnixNano()
+
+		msgTransfer := testsuite.GetMsgTransfer(
+			channelAtoB.PortID,
+			channelAtoB.ChannelID,
+			transfertypes.V2,
+			testvalues.DefaultTransferCoins(chainADenom),
+			chainAAddress,
+			chainBAddress,
+			clienttypes.ZeroHeight(),
+			uint64(inFiveMinutes),
+			"",
+			nil)
+		resp := s.BroadcastMessages(ctx, chainA, chainAWallet, msgTransfer)
+		s.AssertTxSuccess(resp)
+	})
+
+	t.Run("start relayer", func(t *testing.T) {
+		s.StartRelayer(relayer)
+	})
+
+	chainBDenom := transfertypes.NewDenom(chainADenom, transfertypes.NewHop(channelAtoB.Counterparty.PortID, channelAtoB.Counterparty.ChannelID))
+	t.Run("packet has reached B", func(t *testing.T) {
+		s.AssertPacketRelayed(ctx, chainA, channelAtoB.PortID, channelAtoB.ChannelID, 1)
+
+		balance, err := query.Balance(ctx, chainB, chainBAddress, chainBDenom.IBCDenom())
+		s.Require().NoError(err)
+
+		s.Require().Equal(testvalues.IBCTransferAmount, balance.Int64())
+	})
+
+	t.Run("IBC transfer from B (unwind) to C through A", func(t *testing.T) {
+		inFiveMinutes := time.Now().Add(5 * time.Minute).UnixNano()
+
+		forwarding := transfertypes.NewForwarding(
+			true,
+			transfertypes.NewHop(channelAtoB.PortID, channelAtoB.ChannelID),
+			transfertypes.NewHop(channelBtoC.PortID, channelBtoC.ChannelID),
```

Comment:
You are unwinding to A and then routing through B to C, so I think you could just skip B and send directly to C from A.

```suggestion
			transfertypes.NewHop(channelAtoC.PortID, channelAtoC.ChannelID),
```

Reply:
We don't have a channel from A to C (I think our setup only creates A <-> B and B <-> C https://github.com/cosmos/ibc-go/blob/bznein%2F6578%2Ftest2/e2e/testsuite/testsuite.go#L206 ) though I could create it manually if we prefer

Reply:
ah, yes, I see. I think then it's ok to leave it as is.